cibyl/sources/source.py

```diff
@@ -74,35 +74,50 @@ def request_handler(*args, **kwargs):
 class Source(AttrDict):
     """Represents a data provider within a system."""
 
-    def __init__(self, name: str = None, driver: str = None, **kwargs):
+    def __init__(self,
+                 name: str = None,
+                 driver: str = None,
+                 **kwargs: Argument):
         kwargs.setdefault('enabled', True)
         kwargs.setdefault('_setup', False)
+        kwargs.setdefault('_down', False)
         kwargs.setdefault('priority', 0)
 
         super().__init__(name=name, driver=driver, **kwargs)
 
-    def is_setup(self):
+    def is_setup(self) -> bool:
         """Return wether the source has been setup."""
         return self._setup
 
-    def ensure_source_setup(self):
+    def ensure_source_setup(self) -> None:
         """Ensure that setup is called for the source. If setup was previously
         called, do nothing."""
         if not self.is_setup():
             self._setup = True
             self.setup()
 
+    def is_down(self) -> bool:
+        """Return wether the source has been teardown."""
+        return self._down
+
+    def ensure_teardown(self) -> None:
+        """Ensure that teardown is called for the source. If teardown was
+        previously called, do nothing."""
+        if self.is_setup() or not self.is_down():
```

Comment:
Shouldn't this be an `and`. We want to run the teardown if the setup was called before and the teardown was not

Reply:
If there's no a case where the attribute setup is 1 or the attribute down isn't 0 then you're right.

Reply:
I've changed it. Thanks Joan.